### ISSUE TITLE
Install with pipx on 3.12 crashes on startup

### ISSUE THREAD

**Issue by ssteinerx** [nahco314/poetry2rye#1]
$ poetry2rye 
Traceback (most recent call last):
  File "/home/someguy/.local/bin/poetry2rye", line 5, in <module>
    from poetry2rye.main import main
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/poetry2rye/main.py", line 5, in <module>
    from poetry2rye.convert import convert
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/poetry2rye/convert.py", line 11, in <module>
    from poetry2rye.project import BasicDependency
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/poetry2rye/project.py", line 8, in <module>
    import pyproject_parser
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/pyproject_parser/__init__.py", line 34, in <module>
    import dom_toml
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/dom_toml/__init__.py", line 43, in <module>
    from domdf_python_tools.paths import PathPlus
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/domdf_python_tools/paths.py", line 79, in <module>
    from domdf_python_tools.typing import JsonLibrary, PathLike
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/domdf_python_tools/typing.py", line 36, in <module>
    from typing_extensions import Protocol, runtime_checkable
  File "/home/someguy/.local/pipx/venvs/poetry2rye/lib/python3.12/site-packages/poetry/core/_vendor/typing_extensions.py", line 1174, in <module>
    class TypeVar(typing.TypeVar, _DefaultMixin, _root=True):
TypeError: type 'typing.TypeVar' is not an acceptable base type
****

**Comment by ssteinerx** [comment]
This seems to be related to typing_extensions and the many places it gets itself installed through various _vendor directories by various tools/libraries.
Not sure what there is to do about it yet.

**Comment by nahco314** [comment]
I reproduced it in my environment.

Reading the pipx METADATA, it appears that 4.11.0 of typing_extentions is installed, but 4.11.0 does not appear to contain the code to inherit typing.TypeVar as shown in the error message ( https://github.com/python/typing_extensions/blob/d4d929d44bd984350e2d17726362295f588eaace/src/typing_extensions.py#L1174 )

I'm not sure, but it seems to work with Python 3.12 when installed with rye install, so it could be some kind of pipx problem.

**Comment by ssteinerx** [comment]
@nahco314 I beat my head on this for half the day yesterday.  Turns out that there are five different copies of typing_extensions in the project, almost all in the _vendor directories of various projects.

Long story short -- it's _vendor'd because the project has/may have/supposedly has/theoretically might have/maybe has a  dependence on a specific particular version of typing_extensions.py.  So everybody includes one!  Whee!

The kicker is that it *depends on what gets imported first which typing_extensions you get* and, if you get an older one, anything depending on the newer one stops working.  For bonus points, there does not appear to be a __version__ string inside typing_extensions.py so you can't even easily, in any way I could figure, find out which one the import lottery fairy brought you.

How ironic that the module that's supposed to smooth out compatibility issues in the typing hints system of Python is, itself, the source of this type (hahaha) of problems and doesn't appear to conform to the python convention of having an easily inspectable  __version__ string anywhere I could find.

I never did figure out a way around this so I just did the poetry->rye conversion by hand and spent 1000 times longer on it than I should've.

**Comment by nahco314** [comment]
I finally understand the concept of vendoring now.
So it means that some libraries depend on a version of poetry-core that includes the old typing_extentions.py as a vendor file, which causes compatibility issues by import caching?

I'm outside now so I will try some patch after returning.

**Comment by nahco314** [comment]
I updated some libraries, and it looks working at [0.2.1](https://github.com/nahco314/poetry2rye/commit/47dee3b6eb143243555784f02bf61170d4fad149).

I don't really understand it, but I'm hoping this will work.
Thanks for the report!

**Comment by ssteinerx** [comment]
I'm afraid this is one of those "Python Packaging Things" that "someone" will straighten out 'cause it's certainly a bit out of hand.  I'm afraid, after looking at the  typing_extensions.py file, its resolution is far, far above my pay grade and, by that, I mean "who the heck even writes this stuff much less understands all the nuances and interactions?"  Thanks for your efforts, regardless!

**Comment by ssteinerx** [comment]
> I finally understand the concept of vendoring now. So it means that some libraries depend on a version of poetry-core that includes the old typing_extentions.py as a vendor file, which causes compatibility issues by import caching?

That's essentially correct.  If a library author (*or one of the other libraries the library itself  `requires`*) decides that it just can't live without its own special version of typing_extensions.py, it will "vendor" it and it will end up in `_vendor` inside the package.  I didn't try to figure out which libraries were the culprits so I can't speak to whether poetry-core _vendor'd typing_extensions.py.  It wouldn't surprise me in the least.  I'm not a big fan of poetry.  At all.  That's why I'm converting things to rye and how I found you!

Once I saw five of them, and couldn't easily determine what version *any of them* were, I knew there would be trouble.

I did `pipx install` it again, and `poetry2rye --help` worked fine so I think that's progress.

I then ran it in a completely non-python related project and filed a new issue about what happened then...

Next time I have to convert from poetry to rye, I'll come back to see whether I can get this working again.
